clusterloader2/pkg/measurement/common/pod_startup_latency.go

```diff
@@ -0,0 +1,302 @@
+	glog.Infof("Gathering pod startup latency measurement...")
+	if p.informer == nil || p.stopCh == nil {
+		return []measurement.Summary{}, fmt.Errorf("metric %s has not been started", podStartupLatencyMetricName)
+	}
+
+	scheduleLag := make([]time.Duration, 0)
+	startupLag := make([]time.Duration, 0)
+	watchLag := make([]time.Duration, 0)
+	schedToWatchLag := make([]time.Duration, 0)
+	e2eLag := make([]time.Duration, 0)
+
+	close(p.stopCh)
+	if err := p.gatherScheduleTimes(c); err != nil {
+		return []measurement.Summary{}, err
+	}
+	for namespace, createTimes := range p.createTimes {
+		for name, create := range createTimes {
+			sched, ok := getNamespaceDict(p.scheduleTimes, namespace)[name]
+			if !ok {
+				glog.Infof("Failed to find schedule time for %v", name)
+				continue
+			}
+			run, ok := getNamespaceDict(p.runTimes, namespace)[name]
+			if !ok {
+				glog.Infof("Failed to find run time for %v", name)
+				continue
+			}
+			watch, ok := getNamespaceDict(p.watchTimes, namespace)[name]
+			if !ok {
+				glog.Infof("Failed to find watch time for %v", name)
+				continue
+			}
+
+			scheduleLag = append(scheduleLag, sched.Time.Sub(create.Time))
+			startupLag = append(startupLag, run.Time.Sub(sched.Time))
+			watchLag = append(watchLag, watch.Time.Sub(run.Time))
+			schedToWatchLag = append(schedToWatchLag, watch.Time.Sub(sched.Time))
+			e2eLag = append(e2eLag, watch.Time.Sub(create.Time))
+		}
+	}
+
+	sort.Sort(durationSlice(scheduleLag))
+	sort.Sort(durationSlice(startupLag))
+	sort.Sort(durationSlice(watchLag))
+	sort.Sort(durationSlice(schedToWatchLag))
+	sort.Sort(durationSlice(e2eLag))
+
+	podStartupLatency := &podStartupLatency{
+		CreateToScheduleLatency: extractLatencyMetrics(scheduleLag),
+		ScheduleToRunLatency:    extractLatencyMetrics(startupLag),
+		RunToWatchLatency:       extractLatencyMetrics(watchLag),
+		ScheduleToWatchLatency:  extractLatencyMetrics(schedToWatchLag),
+		E2ELatency:              extractLatencyMetrics(e2eLag),
+	}
+	return []measurement.Summary{podStartupLatency}, nil
+}
+
+func (p *podStartupLatencyMeasurement) gatherScheduleTimes(c clientset.Interface) error {
+	selector := fields.Set{
+		"involvedObject.kind": "Pod",
+		"source":              corev1.DefaultSchedulerName,
+	}.AsSelector().String()
+	options := metav1.ListOptions{FieldSelector: selector}
+	schedEvents, err := c.CoreV1().Events(p.namespace).List(options)
+	if err != nil {
+		return err
+	}
+	for _, event := range schedEvents.Items {
+		namespaceDict := getNamespaceDict(p.createTimes, event.InvolvedObject.Namespace)
+		if _, ok := namespaceDict[event.InvolvedObject.Name]; ok {
+			getNamespaceDict(p.scheduleTimes, event.InvolvedObject.Namespace)[event.InvolvedObject.Name] = event.FirstTimestamp
+		}
+	}
+	return nil
+}
+
+func checkPodFunc(p *podStartupLatencyMeasurement) func(interface{}) {
+	p.createTimes = make(map[string]map[string]metav1.Time)
```

Comment:
initialize all of these in constructor instead.

Reply:
Done.